diff -ur -x Makefile.in -x aclocal.m4 -x configure -x 'config.*' -x ltmain.sh -x '*.spec' -x po ORG2/gst-plugins-base/ext/ogg/gstoggmux.c NEW2/gst-plugins-base/ext/ogg/gstoggmux.c
--- ORG2/gst-plugins-base/ext/ogg/gstoggmux.c	2008-12-03 14:06:23.000000000 +0000
+++ NEW2/gst-plugins-base/ext/ogg/gstoggmux.c	2008-12-03 14:45:13.000000000 +0000
@@ -106,7 +106,8 @@
     GST_STATIC_CAPS ("video/x-theora; "
         "audio/x-vorbis; audio/x-flac; audio/x-speex; audio/x-celt; "
         "application/x-ogm-video; application/x-ogm-audio; video/x-dirac; "
-        "video/x-smoke; text/x-cmml, encoded = (boolean) TRUE")
+        "video/x-smoke; text/x-cmml, encoded = (boolean) TRUE; "
+        "application/x-kate")
     );
 
 static void gst_ogg_mux_base_init (gpointer g_class);
