Index: NEW2/gst-plugins-base/ext/ogg/gstoggmux.c
===================================================================
RCS file: /cvs/gstreamer/gst-plugins-base/ext/ogg/gstoggmux.c,v
retrieving revision 1.103
diff -u -r1.103 gstoggmux.c
--- NEW2/gst-plugins-base/ext/ogg/gstoggmux.c	24 Nov 2008 19:18:58 -0000	1.103
+++ NEW2/gst-plugins-base/ext/ogg/gstoggmux.c	17 Jan 2009 15:46:02 -0000
@@ -106,7 +106,8 @@
     GST_STATIC_CAPS ("video/x-theora; "
         "audio/x-vorbis; audio/x-flac; audio/x-speex; audio/x-celt; "
         "application/x-ogm-video; application/x-ogm-audio; video/x-dirac; "
-        "video/x-smoke; text/x-cmml, encoded = (boolean) TRUE")
+        "video/x-smoke; text/x-cmml, encoded = (boolean) TRUE; "
+        "application/x-kate")
     );
 
 static void gst_ogg_mux_base_init (gpointer g_class);
