diff -ru gxmms2-0.7.0/gxmms2src/gxmms2.c gxmms2-0.7.0-with-lyrics/gxmms2src/gxmms2.c
--- gxmms2-0.7.0/gxmms2src/gxmms2.c	2007-11-15 20:22:46.000000000 +0000
+++ gxmms2-0.7.0-with-lyrics/gxmms2src/gxmms2.c	2008-10-25 17:57:19.000000000 +0100
@@ -33,6 +33,7 @@
 #include "eggtrayicon.h"
 #define LABEL_WIDTH 19
 #define TITLE_STR_SIZE 512
+#define LYRICS_STR_SIZE 128
 #define FONT_SIZE 12
 
 static GtkWidget    *label_time, *eventbox_time, *scroller;
@@ -43,6 +44,7 @@
 static GtkTooltips  *docklet_tooltip;
 static gboolean      update_scale = TRUE;
 static gboolean      flag_time_remaining = FALSE;
+static gboolean      flag_lyrics = FALSE;
 xmmsc_connection_t  *connection;
 static trackinfo     track;
 static gchar         title_str[TITLE_STR_SIZE] = "gxmms2 (not connected)";
@@ -50,6 +52,7 @@
 static GtkWidget    *wv;
 static ConfigFile    cf;
 static gboolean      local_xmms2d = FALSE;
+static gchar         current_lyrics[LYRICS_STR_SIZE] = "";
 
 static int xmms2_connect(void);
 
@@ -62,7 +65,13 @@
 {
 	const gchar *current_str = wejpscroller_get_text(WEJPSCROLLER(scroller));
 
-	trackinfo_get_full_title(title_str, TITLE_STR_SIZE, track, FALSE);
+	if (current_lyrics[0]) {
+		strncpy(title_str, current_lyrics, TITLE_STR_SIZE);
+		title_str[TITLE_STR_SIZE-1] = 0;
+	}
+	else {
+		trackinfo_get_full_title(title_str, TITLE_STR_SIZE, track, FALSE);
+	}
 	if (!current_str) {
 		wejpscroller_set_text(WEJPSCROLLER(scroller), title_str, TRUE);
 	} else if (strcmp(current_str, title_str)) {
@@ -134,6 +143,65 @@
 	change_volume(-1, wejpvolume_get_volume(WEJPVOLUME(widget)));
 }
 
+static void n_lyrics_set(xmmsc_result_t *res, void *userdata)
+{
+	const gchar *lyrics;
+	
+	if (flag_lyrics) {
+		if (xmmsc_result_get_string(res, &lyrics) && 
+			xmms2ctrl_get_error() != ERROR_CONNECTION) {
+			if (lyrics) {
+				strncpy(current_lyrics, lyrics, LYRICS_STR_SIZE);
+				current_lyrics[LYRICS_STR_SIZE-1] = 0;
+			}
+			else {
+				strcpy(current_lyrics, "");
+			}
+		}
+		else {
+			strcpy(current_lyrics, "");
+		}
+		update_title_label(title_str, &track, TRUE);
+	}
+
+	xmmsc_result_unref(res);
+}
+
+static void get_current_lyrics()
+{
+	if (flag_lyrics) {
+		xmmsc_result_t *res;
+		res = xmmsc_playback_current_lyrics (connection);
+		xmmsc_result_notifier_set(res, n_lyrics_set, NULL);
+		xmmsc_result_unref(res);
+	}
+}
+
+static void sig_handle_current_lyrics(xmmsc_result_t *res, void *userdata)
+{
+	const char *lyrics = NULL;
+
+	if (flag_lyrics &&
+		xmmsc_result_get_string(res, &lyrics) && 
+		xmms2ctrl_get_error() != ERROR_CONNECTION) {
+
+		if (lyrics) {
+			strncpy(current_lyrics, lyrics, LYRICS_STR_SIZE);
+			current_lyrics[LYRICS_STR_SIZE-1] = 0;
+		}
+		else {
+			strcpy(current_lyrics, "");
+		}
+		update_title_label(title_str, &track, TRUE);
+	}
+
+	xmmsc_result_t *res2;
+	res2 = xmmsc_result_restart(res);
+	xmmsc_result_unref(res2);
+
+	xmmsc_result_unref(res);
+}
+
 static void show_gxmms2_about_dialog(void)
 {
 	GtkWidget *dialog;
@@ -317,6 +385,17 @@
 	}
 }
 
+static void update_lyrics_flag()
+{
+	if (flag_lyrics) {
+		get_current_lyrics();
+	}
+	else {
+		strcpy(current_lyrics, "");
+		update_title_label(title_str, &track, TRUE);
+	}
+}
+
 static gboolean cb_key_press(GtkWidget   *widget,
                              GdkEventKey *event,
                              gpointer     userdata)
@@ -352,6 +431,10 @@
 			case GDK_t:
 				flag_time_remaining = !flag_time_remaining;
 				break;
+			case GDK_l:
+				flag_lyrics = !flag_lyrics;
+				update_lyrics_flag();
+				break;
 			case GDK_0:
 				rate_current_track(1);
 				break;
@@ -419,6 +502,9 @@
 		xmms2ctrl_get_error() != ERROR_CONNECTION) {
 		xmmsc_result_t *res2;
 
+		/* ensure we don't end up with stale lyrics when changing track */
+		strcpy (current_lyrics, "");
+
 		res2 = xmmsc_medialib_get_info(connection, current_id);
 		xmmsc_result_notifier_set(res2, n_media_lib_get_info, &current_id);
 		xmmsc_result_unref(res2);
@@ -431,6 +517,9 @@
 		xmms2ctrl_get_error() != ERROR_CONNECTION) {
 		xmmsc_result_t *res2;
 
+		/* ensure we don't end up with stale lyrics when changing track */
+		strcpy (current_lyrics, "");
+
 		res2 = xmmsc_medialib_get_info(connection, current_id);
 		xmmsc_result_notifier_set(res2, n_media_lib_get_info, &current_id);
 		xmmsc_result_unref(res2);
@@ -589,6 +678,8 @@
 		                  bc_handle_playback_status_change, NULL);
 		XMMS_CALLBACK_SET(connection, xmmsc_broadcast_playback_volume_changed,
 		                  bc_handle_volume_change, NULL);
+		XMMS_CALLBACK_SET(connection, xmmsc_signal_playback_lyrics_changed,
+		                  sig_handle_current_lyrics, connection);
 
 		xmmsc_disconnect_callback_set(connection, connection_lost, NULL);
 		
@@ -599,6 +690,10 @@
 		res = xmmsc_playback_volume_get(connection);
 		xmmsc_result_notifier_set(res, n_volume_init, NULL);
 		xmmsc_result_unref(res);
+
+		/* if we're connecting to a server that is already playing, make sure
+		   we get any current lyrics now rather than wait for it to change */
+		get_current_lyrics();
 	}
 	return xmms2ctrl_get_error();
 }
@@ -718,6 +813,7 @@
 	cfg_add_key(&cf, "ShowRemainingTime",             "no");
 	cfg_add_key(&cf, "TerminateXmms2dOnExit",         "no");
 	cfg_add_key(&cf, "StartLocalXmms2dOnStartup",     "no");
+	cfg_add_key(&cf, "ShowLyrics",                    "yes");
 	username = g_get_user_name();
 	str_length = strlen(username) + 128;
 	ipc_path_default = g_malloc(str_length * sizeof(gchar));
@@ -1110,6 +1206,9 @@
 	if (strncasecmp(cfg_get_key_value(cf, "ShowRemainingTime"), "yes", 3) == 0)
 		flag_time_remaining = TRUE;
 
+	if (strncasecmp(cfg_get_key_value(cf, "ShowLyrics"), "yes", 3) == 0)
+		flag_lyrics = TRUE;
+
 	gtk_widget_show_all(window);
 
 	connection = xmmsc_init("gxmms2");
@@ -1191,6 +1290,7 @@
 		}
 	}
 	cfg_add_key(&cf, "ShowRemainingTime", (flag_time_remaining ? "yes" : "no"));
+	cfg_add_key(&cf, "ShowLyrics", (flag_lyrics ? "yes" : "no"));
 
 	cfg_write_config_file(&cf, path_to_config_file);
 
diff -ru gxmms2-0.7.0/Makefile gxmms2-0.7.0-with-lyrics/Makefile
--- gxmms2-0.7.0/Makefile	2007-09-12 20:19:07.000000000 +0100
+++ gxmms2-0.7.0-with-lyrics/Makefile	2008-07-30 20:44:05.000000000 +0100
@@ -11,7 +11,7 @@
 XMMS2_CFLAGS=$(shell pkg-config xmms2-client xmms2-client-glib --cflags)
 XMMS2_LIB=$(shell pkg-config xmms2-client xmms2-client-glib --libs)
 
-CFLAGS=-O2 -Wall -fPIC $(GTK_CFLAGS) $(XMMS2_CFLAGS) -I.
+CFLAGS=-O2 -Wall -fPIC $(GTK_CFLAGS) $(XMMS2_CFLAGS) -I. -I$(HOME)/include/xmms2
 LIBS=$(GTK_LIB) $(XMMS2_LIB)
 
 KRELLLFLAGS=-shared
diff -ru gxmms2-0.7.0/README gxmms2-0.7.0-with-lyrics/README
--- gxmms2-0.7.0/README	2007-03-24 19:34:59.000000000 +0000
+++ gxmms2-0.7.0-with-lyrics/README	2008-10-25 15:25:08.000000000 +0100
@@ -85,6 +85,7 @@
 - b for next track
 - j for medialib/playlist window
 - t for toggle elapsed/remaining time
+- l for toggle lyrics display
 - +/- to increase/decrease the volume
 - 0..4 to rate the current track where 0 is worst and 4 is best
 
