diff -ruN /mnt/cdrom/svn/xiph/ogg-tools/oggsplit/oggsplit.c oggsplit/oggsplit.c
--- /mnt/cdrom/svn/xiph/ogg-tools/oggsplit/oggsplit.c	2008-01-22 13:42:32.000000000 +0000
+++ oggsplit/oggsplit.c	2008-01-26 12:33:47.000000000 +0000
@@ -42,7 +42,7 @@
 "  a flaw in the tool used to produce the file. If you believe this\n"
 "  is the case, make sure to report it to the author/maintainer of\n"
 "  that tool. Also note that as a consequence, some or all of the\n"
-"  files procuduces by this invocation of OggSplit may also be\n"
+"  files produced by this invocation of OggSplit may also be\n"
 "  invalid Ogg files.\n\n";
 
 const char *optstring="cgno:uVh";
diff -ruN /mnt/cdrom/svn/xiph/ogg-tools/oggsplit/stream.c oggsplit/stream.c
--- /mnt/cdrom/svn/xiph/ogg-tools/oggsplit/stream.c	2008-01-22 13:42:32.000000000 +0000
+++ oggsplit/stream.c	2008-01-26 11:47:44.000000000 +0000
@@ -18,6 +18,7 @@
 const char *magic_theora = "\x80theora";
 const char *magic_speex  = "Speex";
 const char *magic_flac   = "fLaC";
+const char *magic_kate   = "\x80kate\0\0\0\0";
 
 int stream_ctrl_init(stream_ctrl_t *sc)
 {
@@ -57,6 +58,8 @@
     stream->type=STREAM_TYPE_SPEEX;
   else if(strncmp(og->body, magic_flac, strlen(magic_flac))==0)
     stream->type=STREAM_TYPE_FLAC;
+  else if(strncmp(og->body, magic_kate, sizeof(magic_kate)-1)==0)
+    stream->type=STREAM_TYPE_KATE;
   else
     stream->type=STREAM_TYPE_UNKNOWN;
 
@@ -112,6 +115,9 @@
   case STREAM_TYPE_FLAC:
     return "flac";
     break;
+  case STREAM_TYPE_KATE:
+    return "kate";
+    break;
   case STREAM_TYPE_UNKNOWN:
   default:
     return "unknown";
diff -ruN /mnt/cdrom/svn/xiph/ogg-tools/oggsplit/stream.h oggsplit/stream.h
--- /mnt/cdrom/svn/xiph/ogg-tools/oggsplit/stream.h	2008-01-22 13:42:32.000000000 +0000
+++ oggsplit/stream.h	2008-01-26 11:48:00.000000000 +0000
@@ -20,6 +20,7 @@
 #define STREAM_TYPE_THEORA    0x02
 #define STREAM_TYPE_SPEEX     0x03
 #define STREAM_TYPE_FLAC      0x04
+#define STREAM_TYPE_KATE      0x05
 
 typedef struct {
   int    serial;
